### PR TITLE
fix expose multi protocols issue

### PR DESCRIPTION
**Pull request by adohe-zz**
Please refer #23882 for more detail. @bprashanth @bgrant0607 ptal.

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/24090)

### CODE REVIEW

File: pkg/kubectl/cmd/expose.go [hunk 1]
```diff
@@ -198,7 +209,10 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 	}
 
 	// Generate new object
+	fmt.Printf("params here: %v\n", params)
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/expose.go]: I assume these prints were for debugging?

**Comment by bgrant0607** [comment]
Thanks!
cc @kubernetes/kubectl

File: pkg/kubectl/cmd/expose.go (hunk 1 above)
Comment (AdoHe) [inline pkg/kubectl/cmd/expose.go]: sure, I will remove this. Must ignore this :)

File: pkg/kubectl/cmd/expose.go [hunk 2]
```diff
@@ -182,6 +182,17 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 			params["ports"] = strings.Join(ports, ",")
 		}
 	}
+
+	// Always try to derive protocols from the exposed object, as
+	// user may specify UDP protocol
+	protocolsMap, err := f.ProtocolsForObject(inputObject)
```
Comment (kargakis) [inline pkg/kubectl/cmd/expose.go]: Please follow the pattern used above:
```go
if protocol, found := params["protocol"]; found {
   // your code
}
```
This assures that only generated objects that need to use protocol will do.

File: pkg/kubectl/service_test.go [hunk 3]
```diff
@@ -404,6 +404,40 @@ func TestGenerateService(t *testing.T) {
...
 			},
 		},
+		{
+			generator: ServiceGeneratorV2{},
+			params: map[string]interface{}{
+				"selector":  "foo=bar",
+				"name":      "test",
+				"ports":     "80,8080",
+				"protocol":  "TCP",
+				"protocols": "8080/UDP",
```
Comment (kargakis) [inline pkg/kubectl/service_test.go]: add a test with multiple protocols

File: pkg/kubectl/cmd/expose.go (hunk 2 above)
Comment (AdoHe) [inline pkg/kubectl/cmd/expose.go]: @kargakis if we follow this pattern, we have to leave the default value of `protocol` to empty, now it is `TCP` and which just makes expose multi protocols not ok.

Comment (kargakis) [inline pkg/kubectl/cmd/expose.go]: protocol in `params` will be always found for generators that require a protocol such as the service generators. Doesn't have to do anything with the default value in the flag.

File: pkg/kubectl/service.go [hunk 4]
```diff
@@ -134,10 +147,14 @@ func generate(genericParams map[string]interface{}) (runtime.Object, error) {
 		if len(portStringSlice) > 1 {
 			name = fmt.Sprintf("port-%d", i+1)
 		}
+		protocol := params["protocol"]
```
Comment (kargakis) [inline pkg/kubectl/service.go]: What should happen if I expose a service that has multiple ports with different protocols and also use --protocol? It seems that my value will be overriden, right? This is a problem with setting a default in the protocol flag.

Comment (AdoHe) [inline pkg/kubectl/service.go]: IIUC, you mean the priority order, as for now, I will always try to get a protocol from the port protocol map, and if has no such a mapping for a port, I will use the --protocol flag value.

**Comment by AdoHe** [comment]
@kargakis kindly ping.

File: pkg/kubectl/service.go (hunk 4 above)
Comment (kargakis) [inline pkg/kubectl/service.go]: Shouldn't this be the other way around? If I use --protocol I would expect it to override what protocol(s) the resource I am exposing has. Also you need to make sure whether a protocol has been specified with --protocol. Right now, you cannot without removing the default from the --protocol flag.

**Comment by kargakis** [comment]
One comment: https://github.com/kubernetes/kubernetes/pull/24090#discussion_r61242909

**Comment by antonmry** [comment]
@kargakis @AdoHe what's the status of this issue? I didn't realize there is a PR from @AdoHe so I started to work on it but I see it's almost solved now. 

May I help with the missing test or something?. Let me know if it's case, I would like to contribute here

**Comment by kargakis** [comment]
@antonmry this is almost done. @AdoHe can you address my last comment? All you need to do is remove the default from the [--protocol flag](https://github.com/kubernetes/kubernetes/blob/4642621baf4f9675f05a5d2002f84b49859eeeec/pkg/kubectl/cmd/expose.go#L104) - "TCP" should be "". Then, inside the service generator, you should have something like the following:
```go
protocol := params["protocol"]

switch {
   case len(protocol) == 0 && len(portProtocolMap) == 0:
   // Default to TCP, what the flag was doing previously
  protocol = "TCP"
  case len(protocol) > 0 && len(portProtocolMap) > 0:
  // User has specified --protocol while exposing a multiprotocol resource
  // We should stomp multiple protocols with the one specified
  case len(protocol) == 0 && len(portProtocolMap) > 0:
  // no --protocol and we expose a multiprotocol resource
  protocol = "TCP" // have the default so we can stay sane
  if exposeProtocol, found := portProtocolMap[stillPortString]; found {
			protocol = exposeProtocol
 }
// use protocol
```

File: pkg/kubectl/service.go (hunk 4 above)
Comment (kargakis) [inline pkg/kubectl/service.go]: @AdoHe see https://github.com/kubernetes/kubernetes/pull/24090#issuecomment-217890486 for what I have been thinking about.

**Comment by AdoHe** [comment]
@kargakis I will rebase this and update it.

**Comment by AdoHe** [comment]
@antonmry if you have any comments, please leave it here.

**Comment by antonmry** [comment]
@AdoHe no comments, just let me know if I can help with somehting

**Comment by AdoHe** [comment]
@kargakis ptal.

File: pkg/kubectl/cmd/expose.go [hunk 5]
```diff
@@ -101,7 +101,7 @@ func NewCmdExposeService(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	}
 	cmdutil.AddPrinterFlags(cmd)
 	cmd.Flags().String("generator", "service/v2", "The name of the API generator to use. There are 2 generators: 'service/v1' and 'service/v2'. The only difference between them is that service port in v1 is named 'default', while it is left unnamed in v2. Default is 'service/v2'.")
-	cmd.Flags().String("protocol", "TCP", "The network protocol for the service to be created. Default is 'tcp'.")
+	cmd.Flags().String("protocol", "", "The network protocol for the service to be created. Default is 'tcp'.")
```
Comment (kargakis) [inline pkg/kubectl/cmd/expose.go]: Can you capitalize 'tcp' to 'TCP'?

File: pkg/kubectl/cmd/expose.go [hunk 6]
```diff
@@ -198,6 +198,19 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 				params["ports"] = strings.Join(ports, ",")
 			}
 		}
+
+		// Always try to derive protocols from the exposed object, as
+		// user may specify UDP protocol
```
Comment (kargakis) [inline pkg/kubectl/cmd/expose.go]: "may use different protocols for different ports"

**Comment by kargakis** [comment]
A couple of nits, LGTM otherwise

**Comment by AdoHe** [comment]
@kargakis ptal.

**Comment by chancez** [comment]
@kargakis Will this be in 1.3, or can this be considered as a cherry-pick candidate?

**Comment by kargakis** [comment]
> @kargakis Will this be in 1.3, or can this be considered as a cherry-pick candidate?

I think everything in master until tomorrow will be in 1.3 but I am not sure. @bgrant0607 a question for you

**Comment by bgrant0607** [comment]
@kargakis @chancez We haven't cut the 1.3 branch yet.

**Comment by joshperry** [comment]
Does this also address #23880?

**Comment by ensonic** [comment]
Almost a year later I run into this error (my cluster is running 1.5.3). Has this never been merged into a release?

```
The Service "test" is invalid: spec.ports: Invalid value: [{"name":"websocket","protocol":"TCP","port":9090,"targetPort":9090,"nodePort":0},{"name":"vpn","protocol":"UDP","port":1194,"targetPort":1194,"nodePort":0},{"name":"ssh","protocol":"TCP","port":22,"targetPort":22,"nodePort":0}]: cannot create an external load balancer with mix protocols
```

**Comment by adohe** [comment]
@enisoc I believe so, could you please give detailed steps on how to reproduce this? I will take a deep look.

**Comment by enisoc** [comment]
@adohe I think you meant @ensonic.

Hello again, @ensonic. :)

**Comment by kargakis** [comment]
@ensonic what version of kubectl are you using?

**Comment by ensonic** [comment]
```
> kubectl version
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:57:25Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.3", GitCommit:"029c3a408176b55c30846f0faedf56aae5992e9b", GitTreeState:"clean", BuildDate:"2017-02-15T06:34:56Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by kargakis** [comment]
@ensonic your error seems real and unrelated to this PR - you cannot create an external loadbalancer with mixed protocols atm:)

**Comment by adohe** [comment]
@enisoc sorry, my fault :)

**Comment by ensonic** [comment]
Oh, this is https://github.com/kubernetes/kubernetes/issues/23880 then. I hope k8s could do this behind the scenes (create one loadbalancer for each proto transparently, but show them as one).

**Pull request merged**